### DISCUSSION TITLE
Multiple hosts POST_INIT_XX in LTSP.CONF

### DISCUSSION THREAD

**Discussion by Soylor** [ltsp/ltsp#791] [General] [answered]
I'l llike a command line to be executed for all the PC's with specific hostname.

PC hostnames lab-01 to lab-99 execute a POST_INIT_xx command line.

Can it be achieved using LTSP.CONF and a regular expression?

**Comment by alkisg** [comment]
Do you mean this?

```ini
[lab-??]
POST_INIT_CMD="date >/run/date.txt"
```

**Reply by Soylor** [reply]
Yes, but with sections named [pc-01] identifies the section with a pc with the same host name automaticaly?
Do i need and environment variable?

**Comment by alkisg** [comment]
I don't understand your question. Give an example.

**Comment by Soylor** [comment]
Ltsp clients get their names from dhcp, as00, as01…bs00,bs01,bs02...
I want that, xx00** hostname pcs execute a command like you suggested but not as01,as02,bs01,bs02,bs03,bs34,bs105...
A section like this could be?
```
[init/*00]
Commands to execute
```
Rihgt?

**Comment by alkisg** [comment]
This should do it:

```ini
[*00]
POST_INIT_CMD="date >/run/date.txt"
```

**Comment by Soylor** [comment]
I did try

This works for all clients
```
[clients]
  echo $HOSTNAME
  read -p 'Hi: ' abcd
```


This does not work for a EXTERNO3 client
```
[EXTERNO3]
  echo $HOSTNAME
  read -p 'Hi: ' abcd
```

**Comment by alkisg** [comment]
Try with the EXTERNO3 mac address instead, `[xx:xx:xx:xx:xx:xx]`.
Also `read -p` won't work because it's dash, not bash. Use `printf "Hi: "; read abcd`.

**Comment by Soylor** [comment]
It works with read -p and with read and MAC ADDRESS mode

```
[82:F5:6F:C3:6A:41]
  printf "Hi: "; read abcd
```
But it's done 3 times!!! (screenshot)

![image](https://user-images.githubusercontent.com/9199475/217838917-9a99d68c-5b5a-4462-87bd-17a76df84b83.png)

**Comment by Soylor** [comment]
PC hostnames are assigned using DHCP (option host-name "EXTERNO")

Anyway Can it be performed HOSTNAME based? (not using mac address)

Could it be somthing related to #568 ?

**Comment by alkisg** [comment]
It's normal that it's done multiple times. There's PRE_INITRD_BOTTOM_x, POST_INITRD_BOTTOM_x, PRE_INIT_x, POST_INIT_x etc. That's why you're supposed to use `POST_INIT_x=command` instead of plain `command`, to have it executed only once.

Yeah if `[HOSTNAME]` isn't working, something is wrong, for example maybe you're not using the PPA inside or outside your chroot.

**Reply by Soylor** [reply]
PPA is added. It was done with:
```
add-apt-repository ppa:ltsp
apt update
apt upgrade
```
`[HOSTNAME]` is not working.

What else can i check to get it working?

**Comment by alkisg** [comment]
On the client, what is the output of this command?

    ltsp --version

**Comment by Soylor** [comment]
```
ltsp 23.01-1~ubuntu22.04.1
```

**Comment by alkisg** [comment]
Nice. Now try `ip a`, to see your ethernet name, for example `enp2s0`, and then run `sudo /usr/lib/klibc/bin/ipconfig -n enp2s0`, to fake a DHCP request. Upload the output of that command here.

**Comment by Soylor** [comment]
![image](https://user-images.githubusercontent.com/9199475/218080829-a261f1ee-e1f5-4021-8b17-2392093ed6ed.png)

**Comment by alkisg** [comment]
OK, DNS appears to be fine. And what's the output of the following command?

```shell
cat /proc/cmdline /run/net*.conf
```

**Comment by Soylor** [comment]
![image](https://user-images.githubusercontent.com/9199475/218193676-a154169b-b121-46d7-96b4-65601d957c9a.png)

**Comment by alkisg** [comment]
Can you upload your whole ltsp.conf?
Maybe you have another match before [hostname].
A hostname can match only one section, it can't match multiple sections.

**Reply by Soylor** [reply]
This is `ltsp.conf`

```
# /bin/sh -n
# LTSP configuration file
# Documentation=man:ltsp.conf(5)

# The special [server] section is evaluated only by the ltsp server
[server]
# Enable NAT on dual NIC servers
# NAT=1
# Provide a full menu name for x86_32.img when `ltsp ipxe` runs
# IPXE_X86_32_IMG="Debian Bullseye"
 OMIT_IMAGE_EXCLUDES="etc/epoptes/server.key"
 NFS_HOME=1

# The special [common] section is evaluated by both the server and ltsp clients
[common]
# Specify an alternative TFTP_DIR
# TFTP_DIR=/var/lib/tftpboot
 PWMERGE_SUR="siatec|alumno"
# In the special [clients] section, parameters for all clients can be defined.
# Most ltsp.conf parameters should be placed here.
[clients]
# Specify an /etc/fstab line for NFS home; note this is insecure
# FSTAB_HOME="server:/home /home nfs defaults,nolock 0 0"
 FSTAB_HOME="172.19.46.11:/home /home nfs defaults,nolock 0 0"
 DNS_SERVER="172.19.46.12 8.8.8.8"
# epoptes, garantizar servicio arrancado
 KEEP_SYSTEM_SERVICES="epoptes"
 IGNORE_EPOPTES=1
 POST_INIT_AD="/etc/init.d/siteleco"
 POST_INIT_AJUSTES="section_ajustes"
 POST_INIT_SIN_CREAR_HOMES="sudo pam-auth-update --remove mkhomedir"
[EXTERNO3]
  printf "EXTERNO3 Hi: " >>/tmp/test
[ajustes]
 sed -i 's/"1"/"0"/' /etc/apt/apt.conf.d/20auto-upgrades
 sed -i 's/Prompt=.*/Prompt=never/' /etc/update-manager/release-upgrades
```

**Comment by alkisg** [comment]
It looks good to me, except that /tmp/test will be deleted by systemd-tmpfiles so you won't see it at all.
Better use >>/run/test instead of >>/tmp/test.

When I get some free time, I'll try to reproduce your setup and test.

**Comment by Soylor** [comment]
I changed path but not changes. (thanks)

**Comment by Soylor** [comment]
Can i help with trace for getting [hostname] working?

**Comment by alkisg** [comment]
Make a test with a lowercase hostname, e.g. externo instead of EXTERNO3.
If that doesn't work then yeah send a shell trace.
I tested a few things locally and they all worked for me...

**Comment by Soylor** [comment]
Tested with an absolutly empty `ltsp.conf` with only this content:

```
[externo3]
 printf "EXTERNO3 Hixxx: $HOSTNAME" >> /run/test
```
and with only

```
[EXTERNO3]
 printf "XXXEXTERNO3 Hixxx: $HOSTNAME" >> /run/test
 printf "EXTERNO3 Hixxx: $HOSTNAME" >> /run/test
```

No luck, how can i get a shell trace?

**Comment by alkisg** [comment]
Did you change the hostname to lowercase in your DNS server?

Unfortunately documenting how to get a shell trace would require more time than I currently have... :)

**Comment by Soylor** [comment]
I changed it in DHCP server (dhcp server sends hostnames). No DNS is used.

**Comment by Soylor** [comment, marked as answer]
Got it working with lowercase in `dhcpd.conf`
`LTSP.CONF` down not work with UPPERCASE hostnames.

Solved.

Tnakk you
